src/main/java/org/jitsi/videobridge/cc/vp8/VP8AdaptiveTrackProjectionContext.java

```diff
@@ -0,0 +1,458 @@
+        // want to forward. We don't want to be allocating new objects unless
+        // we're interested in the quality of this frame.
+        if (!vp8QualityFilter.acceptFrame(rtpPacket, targetIndex, nowMs))
+        {
+            return null;
+        }
+
+        // We know we want to forward this frame, but we need to make sure it's
+        // going to produce a decodable VP8 packet stream.
+        int maxSequenceNumberSeenBeforeFirstPacket
+            = getMaxSequenceNumberOfFrame(
+                rtpPacket.getSSRCAsLong(), rtpPacket.getTimestamp());
+
+        VP8FrameProjection nextVP8FrameProjection = lastVP8FrameProjection
+            .makeNext(rtpPacket, maxSequenceNumberSeenBeforeFirstPacket, nowMs);
+
+        if (nextVP8FrameProjection == null)
+        {
+            return null;
+        }
+
+        // We have successfully projected the incoming frame and we've allocated
+        // a starting sequence number for it. Any previous frames can no longer
+        // grow.
+        vp8FrameProjectionMap.put(rtpPacket.getTimestamp(), nextVP8FrameProjection);
+        // The frame attached to the "last" projection is no longer the "last".
+        lastVP8FrameProjection = nextVP8FrameProjection;
+
+        // Cleanup the frame projection map.
+        vp8FrameProjectionMap.entrySet().removeIf(
+            e -> e.getValue().isFullyProjected(nowMs));
+
+        return nextVP8FrameProjection;
+    }
+
+    /**
+     * Given a frame (specified by the SSRC and the timestamp that are specified
+     * as arguments), find the highest sequence number we've received from
+     * that frame.
+     *
+     * @param ssrc the SSRC of the frame
+     * @param timestamp the timestamp of the frame
+     * @return the highest sequence number we've received from the frame that is
+     * specified by the SSRC and timestamp arguments.
+     */
+    private int getMaxSequenceNumberOfFrame(long ssrc, long timestamp)
+    {
+        Map<Long, Integer> frameToMaxSequenceNumberMap
+            = ssrcToFrameToMaxSequenceNumberMap.get(ssrc);
+
+        if (frameToMaxSequenceNumberMap == null)
+        {
+            return -1;
+        }
+
+        return frameToMaxSequenceNumberMap
+            .getOrDefault(timestamp, -1);
+    }
+
+    /**
+     * Upon arrival of an RTP packet of a video frame (specified by its SSRC,
+     * its timestamp and its sequence number that are specified as arguments),
+     * elects and store the highest sequence number we've received from that
+     * frame.
+     *
+     * @param ssrc the SSRC of the frame
+     * @param timestamp the timestamp of the frame
+     * @param sequenceNumber the sequence number of the RTP packet,
+     * potentially the highest sequence number that we've received from that
+     * frame.
+     */
+    private void storeMaxSequenceNumberOfFrame(
+        long ssrc, long timestamp, int sequenceNumber)
+    {
+        Map<Long, Integer> frameToMaxSequenceNumberMap
```

Comment:
i think it'd be good to type this as an `LRUCache` instead of a map

Reply:
And I'm not sure we want to forget past values (that's what the LRU does)

Reply:
oh, i was just saying that this type _is_ actually an LRUCache, no?   

```
ssrcToFrameToMaxSequenceNumberMap
            .computeIfAbsent(ssrc, k -> new LRUCache<>(5));
```

Reply:
I misunderstood your initial comment, my last commit should address this now.